### ISSUE TITLE
Pytorch gradients with control flow

### ISSUE THREAD

**Issue by jeff-engineml** [uber/horovod#564]
Environment: Python 2.7, Horovod 0.15.0, Pytorch 0.4.0, CUDA 9.0, Cudnn 7

```python
def calc_loss(pred, gt)
  if some_condition():
    loss = torch.sum(torch.abs(pred - gt))
  else:
    loss = 0.0
  return loss

optimizer = torch.optim.SGD(model.parameters(), lr=0.1)
optimizer = hvd.DistributedOptimizer(optimizer, model.named_parameters())

pred = model(x)
loss = calc_loss(pred, ground_truth)
loss.backward()
optimizer.step()
```

This example hangs if `some_condition` evaluates to `False` for some ranks, and `True` for others. For the ranks that `some_condition` evaluate to `True`, pytorch will send gradients for all weights in model (as expected). For ranks that evaluate to `False`, pytorch will not send **any** gradients. When this happens I see messages like:

```allreduce.model.conv_1.0.weight [missing ranks: 0, 1, 2, 3, 4, 6, 7]```

I can fix this by making the `False` case depend on the input `pred`, but this seems a bit hacky.

```python
def calc_loss(pred, gt)
  if some_condition():
    loss = torch.sum(torch.abs(pred - gt))
  else:
    loss = 0.0 * torch.sum(pred)
  return loss
```

Is there a better way to do this?

**Comment by alsrgv** [comment]
I think we could solve this by forcing allreduce for all model parameters in `step()` that were not being allreduced already in gradient hooks.  Would you like to contribute that change?

**Comment by jeff-engineml** [comment]
I'll take a look and see what I can do
